src/main/java/htsjdk/samtools/util/TestUtil.java

```diff
@@ -73,6 +73,20 @@ public static void recursiveDelete(final File directory) {
     }
 
     /**
+     * Schedule a file or directory to be (recursively) deleted on jvm exit.
+     *
+     * This will silently delete the directory as well as it's contents.
+     * It improves upon {@link TestUtil#recursiveDelete(File)} by deleting directories and files that did not exist at call time.
+     *
+     * It cannot be undone, dir will be deleted unless program is killed.
+     *
+     * @param dir to be deleted
+     */
+    public static void deleteRecursivelyOnExit(final File dir){
```

Comment:
In https://github.com/samtools/htsjdk/pull/910, I implemented a [`deleteOnExit(Path)`](https://github.com/samtools/htsjdk/pull/910/files#diff-828bef0ca8f35f36a26a63b0efac01d1R351) using a shutdown hook too but with  a thread implementation that can be tested.

Maybe you can have a look there, and add a version to `IOUtils` that can be used also outside the testing framework...